core/embed/rust/src/micropython/gc.rs

```diff
@@ -0,0 +1,79 @@
+use core::{
+    alloc::Layout,
+    marker::Unsize,
+    ops::{CoerceUnsized, Deref, DispatchFromDyn},
+    ptr::{self, NonNull},
+};
+
+use super::ffi;
+
+/// A pointer type for values on the garbage-collected heap.
+///
+/// Although a garbage-collected pointer type technically should implement
+/// `Copy` and `Clone`, we avoid doing this until proven necessary.
+pub struct Gc<T: ?Sized>(NonNull<T>);
+
+impl<T: ?Sized + Unsize<U>, U: ?Sized> CoerceUnsized<Gc<U>> for Gc<T> {}
+impl<T: ?Sized + Unsize<U>, U: ?Sized> DispatchFromDyn<Gc<U>> for Gc<T> {}
+
+impl<T> Gc<T> {
+    /// Allocate memory on the heap managed by the MicroPython garbage collector
+    /// and then place `v` into it. `v` will _not_ get its destructor called.
+    pub fn new(v: T) -> Self {
+        let layout = Layout::for_value(&v);
+        // TODO: Assert that `layout.align()` is the same as the GC alignment.
+        // SAFETY:
+        //  - Unfortunately we cannot respect `layout.align()` as MicroPython GC does
+        //    not support custom alignment.
+        //  - `ptr` is guaranteed to stay valid as long as it's reachable from the stack
+        //    or the MicroPython heap.
+        unsafe {
+            let raw = ffi::gc_alloc(layout.size(), 0).cast();
+            ptr::write(raw, v);
+            Self::from_raw(raw)
+        }
+    }
+
+    /// Return a mutable reference to the value.
+    ///
+    /// # Safety
+    ///
+    /// `Gc` values can originate in the MicroPython interpreter, and these can
+    /// be both shared and mutable. Before calling this function, you have to
+    /// ensure that `this` is unique for the whole lifetime of the
+    /// returned mutable reference.
+    pub unsafe fn as_mut(this: &mut Self) -> &mut T {
```

Comment:
This is intentionally not `as_mut(&mut self)`, right? It doesn't work with `let map = unsafe { dict.as_mut() }.map_mut();` in `protobuf/decode.rs` and I'm not sure whether to change it here or there.

Reply:
We need to change this in the Protobuf PR, smart pointer types should usually avoid having `self` methods. `Gc::as_mut(dict)` would be used in the example.

Reply:
Proposed fix: a010d9bd8e9041dd390170c8c9e110e1f056a286